### PR TITLE
[TRB-44977] Pre-Scale Execution Condition Checks for Safe Node Scalin…

### PR DESCRIPTION
**Pull request by libai98**
# Intent

For customers looking to gain insights into cluster capacity for running their workloads, we offer provision and suspend actions as a means to understand real-time demand. By utilizing our analysis, customers can accurately determine the number of nodes required, as opposed to relying solely on thresholds used with cluster autoscalers. Our customers seek to align these actions with their specific definitions of minimum and maximum node numbers, and they require Turbo to discover these definitions and allow them to define policies regarding node suspension and provisioning.

wiki: https://ibm-turbo-wiki.fyre.ibm.com/en/CON/AE/Advanced-Engineering/Cloud-Native-Control/new-page

# Background

This is the second part of the implementation epic Node Suspension and Provision: User defined min/max per Node Pool per Cluster. The First part of the implementation is merged under the PR:https://github.com/turbonomic/kubeturbo/pull/913 

# Testing
******* With default NodeGroupSize **min = 1** and max = 1000
```
I0810 21:35:56.945271       1 kubeturbo_builder.go:1037] Start watching the autoreload config file /etc/kubeturbo/turbo-autoreload.config
I0810 21:35:56.947776       1 kubeturbo_builder.go:1075] Cluster nodePoolSize.min current value is 1
I0810 21:35:56.947785       1 kubeturbo_builder.go:1075] Cluster nodePoolSize.max current value is 1000
E0811 01:10:22.944715       1 action_handler.go:238] Failed to execute action SUSPEND on VIRTUAL_MACHINE [/ip-10-0-214-128.us-west-2.compute.internal]: machine set replicas can't be brought down below the minimum nodes of 1
E0811 01:10:22.944725       1 action_handler.go:183] action execution error: machine set replicas can't be brought down below the minimum nodes of 1
E0811 01:10:22.944737       1 turbo_probe.go:202] Error executing action
```

******* Change NodeGroupSize to **min = 0** and max = 1000 through configMap
```
I0811 04:24:45.625851       1 kubeturbo_builder.go:1075] Cluster nodePoolSize.min current value is 1
I0811 04:24:45.625864       1 kubeturbo_builder.go:1094] Cluster nodePoolSize.min changed from 1 to 0
I0811 04:24:45.625868       1 kubeturbo_builder.go:1075] Cluster nodePoolSize.max current value is 1000
I0811 04:24:46.674173
I0811 05:26:16.685039       1 remote_mediation_client.go:216] [handleServerMessages][ServerRequestEndpoint] : received message with request type Action.
I0811 05:26:16.685073       1 action_handler.go:368] Received an action SUSPEND for entity VIRTUAL_MACHINE [ip-10-0-214-128.us-west-2.compute.internal] in namespace []
I0811 05:26:16.710517       1 machine_scale_executor.go:67] Starting to scale down the machineSet cn-latest-lzdjn-worker-us-west-2c by 1 replica
I0811 05:27:36.825634       1 machine_scale_executor.go:88] Completed scaling down the machineSet cn-latest-lzdjn-worker-us-west-2c by 1 replica
```

# Checklist

These are the items that must be done by the developer and by reviewers before the change is ready to merge. Please ~~strikeout~~ any items that are not applicable, but don't delete them

- [ ] Developer Checks
    - [X] Full build with unit tests and fmt and vet checks
    - [ ] Unit tests added / updated
    - [X] No unlicensed images, no third-party code (such as from StackOverflow)
    - [ ] Integration tests added / updated
    - [X] Manual testing done (and described)
    - [ ] Product sweep run and passed
    - [X] Developer wiki updated (and linked to this description)
- [ ] Reviewer Checks
    - [ ] Merge request description clear and understandable
    - [ ] Developer checklist items complete
    - [ ] Functional code review (how is the code written)
    - [ ] Architectural review (does the code try to do the right thing, in the right way)
    - [ ] Defensive coding (incoming data checked / sanitized, exceptions logged, clear error messages)
    - [ ] No unlicensed images, no third-party code (such as from StackOverflow)
    - [ ] Security review checklist complete.

# Audience

_(@ mention any `review/...` groups or people that should be aware of this merge request)_

### CODE REVIEW

**Comment by libai98** [comment]
Before machineset node suspension:
Screenshot 2023-08-10 at 3 40 00 PM
After action execution:
Screenshot 2023-08-11 at 1 28 03 AM

**Review by amd-ibm — APPROVED** [review]

File: cmd/kubeturbo/app/kubeturbo_builder.go [hunk 1]
```diff
@@ -1022,11 +1022,12 @@ func WatchConfigMap() {
 	viper.SetConfigType("json")
 	viper.SetConfigName(autoReloadConfigFileName)
 	for {
-		verr := viper.ReadInConfig()
-		if verr == nil {
+		if verr := viper.ReadInConfig(); verr == nil {
 			break
 		} else {
-			glog.V(4).Infof("Can't read the autoreload config file %s/%s due to the error: %v, will retry in 3 seconds", autoReloadConfigFilePath, autoReloadConfigFileName, verr)
```
Comment (amd-ibm) [inline cmd/kubeturbo/app/kubeturbo_builder.go:1029]: When the file is not found, keep the existing logging at level `4`. What we want is a log at level `1` when the configuration cannot be parsed, such as when there's a missing comma. So, any error that is not `ConfigFileNotFoundError`.

File: cmd/kubeturbo/app/kubeturbo_builder.go [hunk 2]
```diff
@@ -1085,7 +1086,7 @@ func updateNodePoolConfig(configKey string, currentValue *int, defaultValue int,
 
 	newValue, err := strconv.Atoi(newValueStr)
 	if err != nil || newValue < 0 {
-		glog.Errorf("Invalid %s value: %v specified, using default value %d", configKey, newValue, defaultValue)
+		glog.Errorf("Invalid %s value: %s specified, using default value %d", configKey, newValueStr, defaultValue)
```
[the remark this answers is not in the document]
Comment (libai98) [reply cmd/kubeturbo/app/kubeturbo_builder.go]: Just modified that. Thanks!

File: cmd/kubeturbo/app/kubeturbo_builder.go (hunk 1 above)
Comment (libai98) [reply cmd/kubeturbo/app/kubeturbo_builder.go:1029]: Just modified that. Thanks!

Comment (amd-ibm) [reply cmd/kubeturbo/app/kubeturbo_builder.go:1029]: I think your log statements are reversed. Right?

File: cmd/kubeturbo/app/kubeturbo_builder.go [hunk 3]
```diff
@@ -1022,11 +1022,14 @@ func WatchConfigMap() {
...
-		verr := viper.ReadInConfig()
-		if verr == nil {
+		if verr := viper.ReadInConfig(); verr == nil {
 			break
 		} else {
-			glog.V(4).Infof("Can't read the autoreload config file %s/%s due to the error: %v, will retry in 3 seconds", autoReloadConfigFilePath, autoReloadConfigFileName, verr)
+			if _, ok := verr.(viper.ConfigFileNotFoundError); !ok {
+				glog.V(1).Infof("Can't read the autoreload config file %s/%s due to the error: %v, will retry in 3 seconds", autoReloadConfigFilePath, autoReloadConfigFileName, verr)
+			} else {
+				glog.V(4).Infof("Autoreload config file %s/%s not found", autoReloadConfigFilePath, autoReloadConfigFileName)
```
Comment (amd-ibm) [inline cmd/kubeturbo/app/kubeturbo_builder.go:1031]: The config file would be found here. This error message should indicate that the config was invalid.

**Comment by kevinwangcn** [comment]
Looks good to me. Thanks for the improvement on the logging level.

**Pull request merged**